app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt

```diff
@@ -12,9 +18,178 @@ object ControlFlow {
+        nameResolution: ReferenceMap<Any, NamedNode>,
+        defaultParameterValues: ReferenceMap<Function.Parameter, Variable>,
+        diagnostics: Diagnostics
     ): ReferenceMap<Function, ControlFlowGraph> {
-        return TODO()
+        val controlFlowGraphs = ReferenceHashMap<Function, ControlFlowGraph>()
+
+        fun processFunction(function: Function) {
+            val treeRoots = mutableListOf<IFTNode>()
+            var entryTreeRoot: IFTNode? = null
+            val unconditionalLinks = ReferenceHashMap<IFTNode, IFTNode>()
+            val conditionalTrueLinks = ReferenceHashMap<IFTNode, IFTNode>()
+            val conditionalFalseLinks = ReferenceHashMap<IFTNode, IFTNode>()
+
+            fun link(from: Pair<IFTNode, LinkType>?, to: IFTNode) {
+                if (from != null) {
+                    val links = when (from.second) {
+                        LinkType.UNCONDITIONAL -> unconditionalLinks
+                        LinkType.CONDITIONAL_TRUE -> conditionalTrueLinks
+                        LinkType.CONDITIONAL_FALSE -> conditionalFalseLinks
+                    }
+
+                    links[from.first] = to
+                } else
+                    entryTreeRoot = to
+            }
+
+            fun mapLinkType(list: List<Pair<IFTNode, LinkType>?>, type: LinkType) = list.map { it?.copy(second = type) }
+
+            var last = listOf<Pair<IFTNode, LinkType>?>(null)
+            var breaking: MutableList<Pair<IFTNode, LinkType>?>? = null
+            var continuing: MutableList<Pair<IFTNode, LinkType>?>? = null
+
+            fun processBlock(block: StatementBlock) {
+                fun addExpression(expression: Expression, variable: Variable?): IFTNode? {
+                    val cfg = createGraphForExpression(expression, variable)
+
+                    treeRoots.addAll(cfg.treeRoots)
+                    unconditionalLinks.putAll(cfg.unconditionalLinks)
+                    conditionalTrueLinks.putAll(cfg.conditionalTrueLinks)
+                    conditionalFalseLinks.putAll(cfg.conditionalFalseLinks)
+
+                    val entry = cfg.entryTreeRoot
+
+                    if (entry != null) {
+                        for (node in last)
+                            link(node, entry)
+
+                        last = cfg.finalTreeRoots.map { Pair(it, LinkType.UNCONDITIONAL) }
+                    }
+
+                    return entry
+                }
+
+                for (statement in block) {
+                    if (last.isEmpty())
+                        diagnostics.report(ControlFlowDiagnostic.UnreachableStatement(statement))
+
+                    when (statement) {
+                        is Statement.Evaluation -> addExpression(statement.expression, null)
+
+                        is Statement.VariableDefinition -> {
+                            val variable = statement.variable
+
+                            if (variable.kind != Variable.Kind.CONSTANT && variable.value != null)
```

Comment:
can we have a `CONSTANT` inside a function's body?

Reply:
We can. It would work in the same way as in the global constant, except that the scope would be limited. If constants were only allowed in the global context, they should have had their own variant of `Program.Global`.